core/src/main/java/com/scalar/db/storage/cassandra/Cassandra.java

```diff
@@ -121,11 +119,15 @@ public Scanner scan(Scan scan) throws ExecutionException {
     scan = copyAndSetTargetToIfNot(scan);
     operationChecker.check(scan);
 
-    ResultSet results = handlers.select().handle(scan);
-
-    return new ScannerImpl(
-        results,
-        new ResultInterpreter(scan.getProjections(), metadataManager.getTableMetadata(scan)));
+    ResultInterpreter resultInterpreter =
+        new ResultInterpreter(scan.getProjections(), metadataManager.getTableMetadata(scan));
+    if (scan.getConjunctions().isEmpty()) {
+      return new ScannerImpl(handlers.select().handle(scan), resultInterpreter);
+    } else {
+      // Ignore limit to control it in FilterableScannerImpl
+      Scan scanAll = Scan.newBuilder(scan).limit(0).build();
+      return new FilterableScannerImpl(scan, handlers.select().handle(scanAll), resultInterpreter);
+    }
```

Comment:
I introduce a new filterable scanner implementation so that the regular ScanAll (w/o filtering) does not have to check the existence of filters every time when iterating. It's also nice to keep the logic simple.